src/EFCore.SqlServer/Query/Internal/SqlServerCompiledQueryCacheKeyGenerator.cs

```diff
@@ -0,0 +1,80 @@
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public SqlServerCompiledQueryCacheKeyGenerator(
+            [NotNull] CompiledQueryCacheKeyGeneratorDependencies dependencies,
+            [NotNull] RelationalCompiledQueryCacheKeyGeneratorDependencies relationalDependencies,
+            [NotNull] ISqlServerConnection sqlServerConnection)
+            : base(dependencies, relationalDependencies)
+        {
+            Check.NotNull(sqlServerConnection, nameof(sqlServerConnection));
+
+            _sqlServerConnection = sqlServerConnection;
+        }
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public override object GenerateCacheKey(Expression query, bool async)
+            => new SqlServerCompiledQueryCacheKey(
+                GenerateCacheKeyCore(query, async),
+                _sqlServerConnection.IsMultipleActiveResultSetsEnabled);
+
+        private readonly struct SqlServerCompiledQueryCacheKey
+        {
+            private readonly RelationalCompiledQueryCacheKey _relationalCompiledQueryCacheKey;
+            private readonly bool _multipleActiveResultSetsEnabled;
+
+            public SqlServerCompiledQueryCacheKey(
+                RelationalCompiledQueryCacheKey relationalCompiledQueryCacheKey, bool multipleActiveResultSetsEnabled)
+            {
+                _relationalCompiledQueryCacheKey = relationalCompiledQueryCacheKey;
+                _multipleActiveResultSetsEnabled = multipleActiveResultSetsEnabled;
+            }
+
+            public override bool Equals(object obj)
+                => !(obj is null)
+                   && obj is SqlServerCompiledQueryCacheKey sqlServerCompiledQueryCacheKey
+                   && Equals(sqlServerCompiledQueryCacheKey);
+
+            private bool Equals(SqlServerCompiledQueryCacheKey other)
+                => _relationalCompiledQueryCacheKey.Equals(other._relationalCompiledQueryCacheKey)
+                   && _multipleActiveResultSetsEnabled == other._multipleActiveResultSetsEnabled;
+
+            public override int GetHashCode()
+            {
+                unchecked
+                {
+                    return (_relationalCompiledQueryCacheKey.GetHashCode() * 397) ^ _multipleActiveResultSetsEnabled.GetHashCode();
```

Comment:
```suggestion
                    return (_relationalCompiledQueryCacheKey.GetHashCode() * 3) ^ _multipleActiveResultSetsEnabled.GetHashCode();
```

3 is enough for a bool

Reply:
Why not just use HashCode.Combine?

Reply:
Picked code from 2.2 It did not have HashCode.Combine back then. I will update.